integration/upload_test.go

```diff
@@ -0,0 +1,257 @@
+			if msg.Topic.String() == topicRequest && msg.Path == pathRequest {
+				requests = append(requests, msg.Value)
+				return len(requests) == uploadFilesCount, nil
+			}
+			return false, fmt.Errorf(msgUnexpectedValue, msg.Value)
+		})
+	require.NoError(suite.T(), err, "error processing file upload requests")
+	require.Equal(suite.T(), uploadFilesCount, len(requests), "wrong file upload request events count")
+
+	requestedFiles := make(map[string]string)
+	for _, request := range requests {
+		uploadRequest := &uploadRequest{}
+		err := parseEventValue(request, uploadRequest)
+		require.NoErrorf(suite.T(), err, "cannot convert %v to upload request, will ignore the event", request)
+		require.NotNilf(suite.T(), uploadRequest.Options, "no upload request options found in payload(%v)", uploadRequest)
+		path, ok := uploadRequest.Options[keyFilePath]
+		require.Truef(suite.T(), ok, "%s key not found in upload request event options", keyFilePath)
+		suite.T().Logf("file upload request: %s, with correlation id: %s", path, uploadRequest.CorrelationID)
+		requestedFiles[uploadRequest.CorrelationID] = path
+	}
+	suite.T().Logf("%v file upload requests, initiating uploads", len(requestedFiles))
+	return requestedFiles
+}
+
+func (suite *fileUploadSuite) startUploads(testUpload upload, requestedFiles map[string]string, files []string) {
+	pathLastUpload := util.GetFeaturePropertyPath(featureID, propertyLastUpload)
+	topicModified := util.GetTwinEventTopic(suite.ThingCfg.DeviceID, protocol.ActionModified)
+
+	connEvents, err := util.NewDigitalTwinWSConnection(suite.Cfg)
+	require.NoError(suite.T(), err, msgFailedCreateWebsocketConnection)
+	defer connEvents.Close()
+
+	util.SubscribeForWSMessages(suite.Cfg, connEvents, typeEvents, fmt.Sprintf(eventFilterTemplate, pathLastUpload))
+	fileIDs := make(map[string]string)
+	for startID, path := range requestedFiles {
+		fileIDs[path] = startID
+		_, err := util.ExecuteOperation(suite.Cfg, suite.featureURL, operationStart, testUpload.requestUpload(startID, path))
+		require.NoErrorf(suite.T(), err, msgErrorExecutingOperation, operationStart)
+	}
+
+	statuses := []interface{}{}
+	err = util.ProcessWSMessages(suite.Cfg, connEvents,
+		func(msg *protocol.Envelope) (bool, error) {
+			if msg.Topic.String() == topicModified && msg.Path == pathLastUpload {
+				statuses = append(statuses, msg.Value)
+				return isTerminal(msg.Value), nil
+			}
+			return false, fmt.Errorf(msgUnexpectedValue, msg.Value)
+		})
+	require.NoError(suite.T(), err, "error processing upload status events")
+
+	lastUploadProgress := 0
+	for ind, status := range statuses {
+		uploadStatus := uploadStatus{}
+		err := parseEventValue(status, &uploadStatus)
+		require.NoErrorf(suite.T(), err, "cannot convert %v to upload request, will ignore the event", status)
+		suite.T().Logf("upload status event(%v)", uploadStatus)
+		require.True(suite.T(), uploadStatus.Progress >= lastUploadProgress && uploadStatus.Progress <= 100,
```

Comment:
It's more correct to use two assert statements here: LessOrEqual, GreaterOrEqual.